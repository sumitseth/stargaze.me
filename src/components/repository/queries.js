import gql from 'graphql-tag';

export const searchReposQuery = gql`
  query($query: String!, $after: String) {
    search(type: REPOSITORY, query: $query, first: 25, after: $after) {
      repositoryCount
      pageInfo {
        endCursor
        hasNextPage
      }
      nodes {
        ... on Repository {
          id
          name
          url
          description
          viewerHasStarred
        }
      }
    }
  }
`;

export const starredReposQuery = gql`
  query StarredReposQuery($numRepos: Int, $after: String) {
    viewer {
      id
      name
      avatarUrl
      starredRepositories(first: $numRepos, after: $after) {
        totalCount
        pageInfo {
          endCursor
          hasNextPage
        }
        nodes {
          id
          name
          description
          url
          viewerHasStarred
        }
      }
    }
  }
`;
