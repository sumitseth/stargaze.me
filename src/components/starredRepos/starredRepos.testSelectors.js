export const starredReposTestSelectors = {
  header: 'starredRepos-header',
  loading: 'starredRepos-loading',
  error: 'starredRepos-error',
  loadMoreButton: 'starredRepos-loadMore-button',
};
