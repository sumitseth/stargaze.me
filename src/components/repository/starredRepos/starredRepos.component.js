import React from 'react';
import PropTypes from 'prop-types';
import { useMutation, useQuery } from '@apollo/react-hooks';
import { Box, Button } from 'rebass';

import { RepositoriesList } from '../repositoryList/repositoriesList.component';
import { starredReposTestSelectors } from './starredRepos.testSelectors';
import { starredReposQuery } from '../queries';
import {
  addStarMutation,
  removeStarMutation,
  addStarAction,
  removeStarAction,
} from '../mutation';

const updateRepositories = (prevResult = {}, { fetchMoreResult = {} }) => {
  const previousViewer = prevResult.viewer || {};
  const currentViewer = fetchMoreResult.viewer || {};
  const previousNodes = previousViewer.starredRepositories.nodes || [];
  const currentNodes = currentViewer.starredRepositories.nodes || [];
  const pageInfo = fetchMoreResult.viewer.starredRepositories.pageInfo;
  return currentNodes.length
    ? {
        ...prevResult,
        viewer: {
          ...previousViewer,
          starredRepositories: {
            ...previousViewer.starredRepositories,
            nodes: [...previousNodes, ...currentNodes],
            pageInfo,
          },
        },
      }
    : prevResult;
};

export const StarredRepos = ({ numRepos = 25 }) => {
  const { loading, error, data, fetchMore } = useQuery(starredReposQuery, {
    variables: { numRepos: numRepos },
  });

  const [addStar] = useMutation(addStarMutation);
  const [removeStar] = useMutation(removeStarMutation);

  if (error)
    return (
      <Box mt={2} data-testid={starredReposTestSelectors.error}>
        There was an error loading repositories!
      </Box>
    );
  if (loading)
    return (
      <Box mt={2} data-testid={starredReposTestSelectors.loading}>
        Loading...
      </Box>
    );
  const hasMoreRepos =
    data.viewer && data.viewer.starredRepositories.pageInfo.hasNextPage;

  return (
    <div>
      <header>
        <hgroup>
          <h2>Your starred Repos</h2>
          <h4 data-testid={starredReposTestSelectors.header}>
            You have starred {data.viewer.starredRepositories.totalCount} repos
          </h4>
        </hgroup>
      </header>
      <hr />
      <RepositoriesList
        repositories={data.viewer.starredRepositories}
        onAddStar={id => addStar(addStarAction(id))}
        onRemoveStar={id => removeStar(removeStarAction(id))}
      />
      {hasMoreRepos ? (
        <Button
          style={{
            marginBottom: '1em',
          }}
          data-testid={starredReposTestSelectors.loadMoreButton}
          onClick={() => {
            fetchMore({
              starredReposQuery,
              variables: {
                numRepos: numRepos || 25,
                after: data.viewer.starredRepositories.pageInfo.endCursor,
              },
              updateQuery: updateRepositories,
            });
          }}
        >
          Show more
        </Button>
      ) : null}
    </div>
  );
};

StarredRepos.propTypes = {
  numRepos: PropTypes.number,
};

export default StarredRepos;
