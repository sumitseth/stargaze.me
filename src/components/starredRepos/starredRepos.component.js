import React from 'react';
import PropTypes from 'prop-types';
import { useQuery } from '@apollo/react-hooks';
import gql from 'graphql-tag';

import { RepositoriesList } from '../repository/repositoriesList.component';
import { starredReposTestSelectors } from './starredRepos.testSelectors';

export const starredReposQuery = gql`
  query StarredReposQuery($numRepos: Int, $after: String) {
    viewer {
      id
      name
      avatarUrl
      starredRepositories(first: $numRepos, after: $after) {
        totalCount
        pageInfo {
          endCursor
          hasNextPage
        }
        nodes {
          id
          name
          description
          url
        }
      }
    }
  }
`;

const updateRepositories = (prevResult = {}, { fetchMoreResult = {} }) => {
  const previousViewer = prevResult.viewer || {};
  const currentViewer = fetchMoreResult.viewer || {};
  const previousNodes = previousViewer.starredRepositories.nodes || [];
  const currentNodes = currentViewer.starredRepositories.nodes || [];
  const pageInfo = fetchMoreResult.viewer.starredRepositories.pageInfo;
  return currentNodes.length
    ? {
        ...prevResult,
        viewer: {
          ...previousViewer,
          starredRepositories: {
            ...previousViewer.starredRepositories,
            nodes: [...previousNodes, ...currentNodes],
            pageInfo,
          },
        },
      }
    : prevResult;
};

export const StarredRepos = ({ numRepos = 25 }) => {
  const { loading, error, data, fetchMore } = useQuery(starredReposQuery, {
    variables: { numRepos: numRepos },
  });

  if (error)
    return (
      <div data-testid={starredReposTestSelectors.error}>
        There was an error loading repositories!
      </div>
    );
  if (loading)
    return (
      <div data-testid={starredReposTestSelectors.loading}>Loading...</div>
    );
  const hasMoreRepos =
    data.viewer && data.viewer.starredRepositories.pageInfo.hasNextPage;

  return (
    <div>
      <header>
        <hgroup>
          <h2>Your starred Repos</h2>
          <h4 data-testid={starredReposTestSelectors.header}>
            You have starred {data.viewer.starredRepositories.totalCount} repos
          </h4>
        </hgroup>
      </header>
      <hr />
      <RepositoriesList repositories={data.viewer.starredRepositories} />
      {hasMoreRepos ? (
        <button
          data-testid={starredReposTestSelectors.loadMoreButton}
          onClick={() => {
            fetchMore({
              starredReposQuery,
              variables: {
                numRepos: numRepos || 25,
                after: data.viewer.starredRepositories.pageInfo.endCursor,
              },
              updateQuery: updateRepositories,
            });
          }}
        >
          Show more
        </button>
      ) : null}
    </div>
  );
};

StarredRepos.propTypes = {
  numRepos: PropTypes.number,
};

export default StarredRepos;
