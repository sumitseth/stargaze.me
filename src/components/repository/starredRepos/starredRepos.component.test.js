import React from 'react';
import { GraphQLError } from 'graphql';
import { MockedProvider } from '@apollo/react-testing';
import { render, getByTestId, waitForElement } from '@testing-library/react';
import StarredRepos from './starredRepos.component';
import { starredReposTestSelectors } from './starredRepos.testSelectors';
import { starredReposQuery } from '../queries';

const STARRED_REPOS_RESPONSE = {
  viewer: {
    id: 'MDQ6VXNlcjMyODgwNQ==',
    name: 'Theofanis Despoudis',
    avatarUrl: 'https://avatars0.githubusercontent.com/u/328805?v=4',
    starredRepositories: {
      totalCount: 100,
      pageInfo: {
        endCursor: 'Y3Vyc29yOnYyOpHOAWbWew==',
        hasNextPage: true,
        __typename: 'pageInfo',
      },
      nodes: [
        {
          id: 'MDEwOlJlcG9zaXRvcnkyMTM0OTI2Ng==',
          name: 'computer-science-for-javascript',
          description:
            'A complete collection of useful data structures and algorithms in JavaScript',
          url: 'https://github.com/theodesp/computer-science-for-javascript',
          viewerHasStarred: true,
          __typename: 'Repo',
        },
      ],
      __typename: 'StarredRepo',
    },
    __typename: 'Viewer',
  },
};

const STARRED_REPOS_MOCKS = [
  {
    request: {
      query: starredReposQuery,
      variables: { numRepos: 1 },
    },
    result: { data: STARRED_REPOS_RESPONSE },
  },
];

const STARRED_REPOS_ERROR = [
  {
    request: {
      query: starredReposQuery,
      variables: { numRepos: 1 },
    },
    result: {
      errors: [new GraphQLError('Error')],
    },
  },
];

describe('<StarredRepos />', () => {
  describe('Loading State', () => {
    let rendered;
    beforeEach(() => {
      rendered = render(
        <MockedProvider mocks={STARRED_REPOS_MOCKS}>
          <StarredRepos numRepos={1} />
        </MockedProvider>
      );
    });
    it('renders a loading state', async () => {
      const { container } = rendered;
      const loading = getByTestId(container, starredReposTestSelectors.loading);
      expect(loading.textContent).toBe('Loading...');
    });

    it('matches loading snapshot', async () => {
      const { container } = rendered;
      expect(container).toMatchSnapshot();
    });
  });

  describe('Loaded State', () => {
    let rendered;
    beforeEach(() => {
      rendered = render(
        <MockedProvider mocks={STARRED_REPOS_MOCKS}>
          <StarredRepos numRepos={1} />
        </MockedProvider>
      );
    });
    it('renders a simple query properly', async () => {
      const { container } = rendered;

      const header = await waitForElement(
        () => getByTestId(container, starredReposTestSelectors.header),
        { container }
      );
      expect(header.textContent).toBe('You have starred 100 repos');

      const loadMoreButton = getByTestId(
        container,
        starredReposTestSelectors.loadMoreButton
      );
      expect(loadMoreButton).toBeVisible();
      expect(loadMoreButton.textContent).toBe('Show more');
    });

    it('matches loaded snapshot', async () => {
      const { container } = rendered;
      await waitForElement(
        () => getByTestId(container, starredReposTestSelectors.header),
        { container }
      );
      expect(container).toMatchSnapshot();
    });
  });

  describe('Error State', () => {
    let rendered;
    beforeEach(() => {
      rendered = render(
        <MockedProvider mocks={STARRED_REPOS_ERROR}>
          <StarredRepos numRepos={1} />
        </MockedProvider>
      );
    });
    it('renders an error state', async () => {
      const { container } = rendered;
      const error = await waitForElement(
        () => getByTestId(container, starredReposTestSelectors.error),
        { container }
      );
      expect(error.textContent).toBe(
        'There was an error loading repositories!'
      );
    });

    it('matches error snapshot', async () => {
      const { container } = rendered;
      await waitForElement(
        () => getByTestId(container, starredReposTestSelectors.error),
        { container }
      );
      expect(container).toMatchSnapshot();
    });
  });
});
