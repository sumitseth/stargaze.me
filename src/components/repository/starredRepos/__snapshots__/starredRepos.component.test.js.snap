// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StarredRepos /> Error State matches error snapshot 1`] = `
<div>
  <div
    class="sc-bdVaJa kmtHOY"
    data-testid="starredRepos-error"
  >
    There was an error loading repositories!
  </div>
</div>
`;

exports[`<StarredRepos /> Loaded State matches loaded snapshot 1`] = `
<div>
  <div>
    <header>
      <hgroup>
        <h2>
          Your starred Repos
        </h2>
        <h4
          data-testid="starredRepos-header"
        >
          You have starred 
          100
           repos
        </h4>
      </hgroup>
    </header>
    <hr />
    <ul
      class="sc-htoDjs htrlDF"
      data-testid="repositories-list"
    >
      <div
        class="sc-bdVaJa sc-gzVnrw beCVhZ"
      >
        <img
          alt="viewerHasStarred"
          src="[object Object]"
          style="vertical-align: middle; cursor: pointer;"
        />
        <h2
          data-testid="repository-name"
        >
          computer-science-for-javascript
        </h2>
        <a
          data-testid="repository-url"
          href="https://github.com/theodesp/computer-science-for-javascript"
          rel="noopener noreferrer"
          target="_blank"
        >
          https://github.com/theodesp/computer-science-for-javascript
        </a>
        <p
          data-testid="repository-description"
        >
          A complete collection of useful data structures and algorithms in JavaScript
        </p>
      </div>
    </ul>
    <button
      class="sc-bdVaJa sc-EHOje fuetwF"
      color="white"
      data-testid="starredRepos-loadMore-button"
      font-size="inherit"
      font-weight="bold"
      style="margin-bottom: 1em;"
    >
      Show more
    </button>
  </div>
</div>
`;

exports[`<StarredRepos /> Loading State matches loading snapshot 1`] = `
<div>
  <div
    class="sc-bdVaJa kmtHOY"
    data-testid="starredRepos-loading"
  >
    Loading...
  </div>
</div>
`;
